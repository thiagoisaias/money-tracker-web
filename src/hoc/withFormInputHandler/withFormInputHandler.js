import React, { Component } from "react";

/*
Form State - Example
state = {
  isFormValid: false,
  formFields: {
    fieldName: {
      elementType: "input",
      elementConfig: {
        type: "text",
        placeholder: "Enter field placeholder"
      },
      name: "Field Name",
      value: this.props.fieldName || "",
      validation: {
        required: true
      },
      isValid: false,
      touched: false
    }
  }
}
*/

const withFormInputHandler = (WrappedComponent, props) => {
  return class extends Component {
    handleInputChange = (event, formFields, checkFieldValidity) => {
      const target = event.target;
      const value = target.value;
      const name = target.name;

      const currentFormField = formFields[name];

      const updatedField = {
        ...currentFormField,
        value: value,
        isValid: checkFieldValidity(value, currentFormField.validation),
        touched: true
      };
      const updatedFormFields = {
        ...formFields,
        [name]: updatedField
      };

      // Verify if all fields are valid and then if the form is valid
      let isFormValid = true;
      for (let fieldKey in updatedFormFields) {
        isFormValid =
          checkFieldValidity(
            updatedFormFields[fieldKey].value,
            updatedFormFields[fieldKey].validation
          ) && isFormValid;
      }

      return {
        isFormValid: isFormValid,
        formFields: {
          ...updatedFormFields
        }
      };
    };

    render() {
      return (
        <WrappedComponent
          {...this.props}
          handleInputChange={this.handleInputChange}
        />
      );
    }
  };
};

export default withFormInputHandler;
