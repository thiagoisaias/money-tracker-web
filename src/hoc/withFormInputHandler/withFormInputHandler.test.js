import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import withFormInputHandler from "./withFormInputHandler"

describe("withFormInputHandler", () => {
  const component = shallow(<withFormInputHandler />);

  it("renders properly", () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
