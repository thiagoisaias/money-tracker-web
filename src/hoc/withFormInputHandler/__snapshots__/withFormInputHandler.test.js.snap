// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFormInputHandler renders properly 1`] = `<withFormInputHandler />`;
